### PR TITLE
Addressing #50, optimization of tohex

### PR DESCRIPTION
**Pull request by hugmyndakassi**
- This optimizes tohex() a fair bit
- Also fixes the link I missed in November to the old osslsignicode
  project page in the README

(... and removes the abandoned claim regarding osslsigncode, as the last commits were rather recently)

### CODE REVIEW

**Comment by hugmyndakassi** [comment]
Actually I didn't get the unit tests to run as of yet. The compiler gave me quite a few errors when trying with CMake and the `-DBUILD_TESTS=1` which was suggested in the README. I need to review how to get this to build successfully so I can run the test suite locally, too.

Either during, but more likely after the weekend.

**Comment by hugmyndakassi** [comment]
@woodruffw is there a tag for the version of `pe-parse` you use for tests or do you always use the tip of `master`?

**Comment by woodruffw** [comment]
I always use the `pe-parse` that's packaged via `vcpkg`, which should be the most recently released tag (1.3.0).

File: README.md [hunk 1]
```diff
@@ -27,8 +27,8 @@ Other available solutions are deficient:
 
 * WINE implements most of `Wintrust`, but is a massive (and arguably non-native) dependency
 for a single task.
-* [`osslsigncode`](https://sourceforge.net/projects/osslsigncode/) can add signatures and check
-timestamps, but is long-abandoned and CLI-focused.
+* [`osslsigncode`](https://github.com/mtrojnar/osslsigncode) can add signatures and check
```
Comment (woodruffw) [inline README.md:30]: Hmm -- this change is already on `master`. Is your local branch stale?

Comment (exoosh) [reply README.md:30]: Now I wonder.

Comment (exoosh) [reply README.md:30]: Hmm, no it's not on `master` in your public repo at least. [See here](https://github.com/trailofbits/uthenticode/blame/master/README.md#L30).

Comment (hugmyndakassi) [reply README.md:30]: The README on your public master still has the old link. The fix from November only addressed the one at the bottom of the README.

Comment (woodruffw) [reply README.md:30]: Whoops, my mistake. Thanks for updating this one as wlel.

**Comment by woodruffw** [comment]
Looks like the unit tests need to be updated to handle the case change in `tohex`. Could you do that?

**Comment by hugmyndakassi** [comment]
Will do (on it right now), but why does the other PR fail? The change in the other one is even more trivial.

**Comment by hugmyndakassi** [comment]
This is odd, the difference isn't just casing. Some of the hashes appear to have been truncated.

```
(gdb) n
1709      if (String::CaseInsensitiveCStringEquals(lhs, rhs)) {
(gdb) p (char*)rhs
$3 = 0x555555611780 "6663dd7c24fa84fce7f16eb2689952c06cfa22"
(gdb) p (char*)lhs
$4 = 0x6040000033d0 "6663DD7C24FA84FCE7F16E0B02689952C06CFA22"
```

Consider the above. It's missing a piece. Making it more obvious:

```
6663DD7C24FA84FCE7F16E0B02689952C06CFA22
6663dd7c24fa84fce7f16e  b2689952c06cfa22
```

Going through those instances just now. I wonder when that happened and on whose side. I didn't even touch those files until you asked.

**Comment by hugmyndakassi** [comment]
I'm starting to think there was an undetected issue in the `tohex` code until now. what do you think @woodruffw ?

**Comment by woodruffw** [comment]
Yikes, that's concerning. I wonder how that made it though -- I vaguely remember copy-pasting the hashes from `signtool` during testing as a sanity check, but maybe I didn't.

It's strange that only some of the `0`s are disappearing -- I'm not sure what would cause that.

**Comment by hugmyndakassi** [comment]
Blame [say's it's old](https://github.com/trailofbits/uthenticode/blame/master/test/signeddata-test.cpp#L21). And if you copy it from there you'll notice that the size doesn't match what would be expected of a SHA1 in string representation.

**Comment by woodruffw** [comment]
Ugh, I bet I just copy-pasted it from `svcli`'s output and visually checked it instead of confirming the exact string. That's a terrible bug on my part.

Your changes LGTM! Once those tests are fixed (including my terrible bug, sorry again), we can merge.

**Comment by hugmyndakassi** [comment]
I listed the affected hashes in the commit messages (it includes duplicates, because some were used in more than one place).

I also took the liberty to prevent similar future issues by checking against the length of the hash string. Potentially this should/could be implemented in a nicer way by retrieving the expected length based on the hash type.

**Review by woodruffw — APPROVED** [review]

**Pull request merged**